### Manhattanizer
#
#This script is a simplification of the "burrito" wrapper in the GWAS pipeline from Masalia et al 2018 (PLOS)
# 
#

### Read in environments and traits to run through the EMMAX GWAS algorithm

envs<-as.character(read.table("Environments.list")[,1])
traits<-as.character(read.table("Traits.list")[,1])


prefix <- "XRQ_264"


#### List SAM genotypes contained in the SNP set (264 set or 239 set)

SAM<-read.table("Software/XRQ_264.tfam")[,1]



### Run EMMAX in a loop across all trait+environment combinations


for (q in 1:length(envs)) {
  
  print(envs[q])
  
  for (i in 1:length(traits)) {
    
    datafile <- paste(traits[i],"_",envs[q],".txt",sep="")
    print(datafile)

Pheno_in<-read.table(paste("Phenotype data/",datafile,sep=""),header=TRUE)

Pheno_in[Pheno_in=="Na"]<-"NA"

PhenFile<-Pheno_in[Pheno_in$Line%in%SAM,] ### limit phenotypic data to just the genotypes in the SNP set



	write.table(data.frame(PhenFile[,1],PhenFile[,1], PhenFile[,2]),paste("Software/",traits[i],"_",envs[q],".pheno",sep=""),row.names = FALSE,col.names = FALSE, quote = FALSE)


	  EMMAX.pk<-paste("./Software/emmax-intel64 -v -d 10 -t ",paste("./Software/",prefix,sep=""), " -p ", paste("./Software/",traits[i],"_",envs[q],".pheno",sep=""), " -c ", paste("./Software/",prefix,".PCA_EV",sep=""), " -k ", paste("./Software/",prefix,".aIBS.kinf",sep=""), " -o ", paste(traits[i],"_",envs[q],sep=""), sep="")
	  system (EMMAX.pk)

  
### move usefull files generated by emmax and in the loop	
	move<-paste("mv ./",paste(traits[i],"_",envs[q],".ps",sep=""),paste(" ./Tables/PSfiles/",sep=""),sep="")
  system(move)

### delete unnecesary files
system(paste("rm ",paste(traits[i],"_",envs[q],sep=""),".log",sep=""))
system(paste("rm ",paste(traits[i],"_",envs[q],sep=""),".reml",sep=""))
system(paste("rm Software/",paste(traits[i],"_",envs[q],sep=""),".pheno",sep=""))

}
}



